_source/_posts/2022-02-21-build-deploy-node-app-heroku.md

```diff
@@ -0,0 +1,335 @@
+});
+
+module.exports = router;
+```
+
+Here you added a property for the `index.jade` template to use called `loggedIn`.  This property will help the view decide if it should display a `login` or `logout` button.  The rest of the home page will be static.
+
+Open your `routes/users.js` file and replace the contents of the file with the following code:
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+function userRoutes(options) {
+  const oidc = options.oidc;
+
+  router.get("/", oidc.ensureAuthenticated(), function (req, res, next) {
+    res.render("users/index", {
+      loggedIn: true,
+      title: "Express",
+      user: req.userContext.userinfo,
+    });
+  });
+
+  return router;
+}
+
+module.exports.userRoutes = userRoutes;
+```
+
+This file is where the magic from the Okta OIDC middleware shines.  You are passing in the `oidc.ensureAuthenticated` middleware in your route definition.  This function will reroute unauthenticated users to the login route you defined in your `app.js` file.  Because you didn't explicitly set this parameter, it will direct the user to `/login` and trigger the authentication flow.  The model is passed to the view here also contains some information about the user that you will display when they land on this page. 
+
+### Write your frontend code
+
+Now you can write some client code.  By default, the express-generator sets up some views using `jade`.  Jade is a templating engine for Node.js.  The language incorporates some conditionals and flow controls to make templating HTML from your model easier.  The syntax is pretty simple to learn, and there are many HTML to jade converters out there.
+
+Open the `layout.jade` file that was generated with your project and replace the code with the following:
+
+```jade
+doctype html
+html
+  head
+    title= title
+    link(rel='stylesheet' href='/stylesheets/style.css')
+    link(rel='stylesheet' href='css/bootstrap.min.css')
+    script(language='javascript' src='js/bootstrap.min.js')
+
+  body
+    .container
+      header.d-flex.flex-wrap.justify-content-center.py-3.mb-4.border-bottom
+        h5.d-flex.align-items-center.mb-3.mb-md-0.me-md-auto.text-dark.text-decoration-none 
+          a(href="/") Home
+        ul.nav.nav-pills
+          if loggedIn
+            li.nav-item 
+              a(href="users").nav-link Profile
+          li.nav-item
+            if loggedIn
+              form(action="logout" method="post")
+                button(type="submit").nav-link.active Logout
+            else 
+              a(href="Users").nav-link.active Login
+
+      block content
+```
+
+We have used the HTML elements and the classes are chained with a `.` between them. The attributes are bounded by parentheses.  This simple layout takes the `loggedIn` value from the server and determines if it should display a `login` or `logout` button.  It also conditionally shows a tab for the `Profile` page if the user is logged in. 
```

Comment:
Could we clarify this? 
"We have used the HTML elements and the classes are chained with a `.` between them." - i can't tell if these two parts are related, or how they relate. 
Is it: "We have used the HTML elements and the classes chained with a `.` between them."  
OR is it better like this (somewhat separate clauses): "We have used the HTML elements; the classes are chained with a `.` between them."